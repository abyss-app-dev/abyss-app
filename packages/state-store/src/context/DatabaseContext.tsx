import type { SQliteClient } from '@abyss/records';
import { createContext, type ReactNode, useContext } from 'react';

interface DatabaseContextType {
    database: SQliteClient | null;
}

const DatabaseContext = createContext<DatabaseContextType>({ database: null });

interface DatabaseProviderProps {
    database: SQliteClient;
    children: ReactNode;
}

export function DatabaseProvider({ database, children }: DatabaseProviderProps) {
    return <DatabaseContext.Provider value={{ database }}>{children}</DatabaseContext.Provider>;
}

export function useDatabase(): SQliteClient {
    const context = useContext(DatabaseContext);
    if (!context.database) {
        throw new Error('useDatabase must be used within a DatabaseProvider');
    }
    return context.database;
}
