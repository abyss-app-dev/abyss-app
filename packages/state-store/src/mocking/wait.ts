export function wait(ms = 50) {
    return new Promise(resolve => setTimeout(resolve, ms));
}
