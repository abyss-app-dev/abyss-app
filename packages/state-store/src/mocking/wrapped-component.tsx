import type { StateRecorder } from './state-recorder';

export function TestRecorder<T>({ functionToTest, recorder }: { functionToTest: () => T; recorder: StateRecorder }) {
    recorder.record(functionToTest());
    return <>content</>;
}
