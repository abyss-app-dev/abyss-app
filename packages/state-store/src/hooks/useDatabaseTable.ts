import type { SqliteTables } from '@abyss/records';
import { useEffect } from 'react';
import { useDatabase } from '../context/DatabaseContext';
import { useQuery } from './useQuery';

export function useDatabaseTable<T extends keyof SqliteTables>(table: T) {
    const database = useDatabase();

    const query = useQuery(async () => {
        return database.tables[table].list();
    }, [table]);

    useEffect(() => {
        return database.tables[table].subscribe(() => {
            query.refetch();
        });
    }, [database, table, query.setData]);

    return query;
}
