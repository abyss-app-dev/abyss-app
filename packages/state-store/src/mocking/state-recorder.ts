import { render } from '@testing-library/react';
import type { ReactNode } from 'react';
import { expect } from 'vitest';
import { wait } from './wait';

export class StateRecorder {
    private states: unknown[] = [];

    record(state: unknown): void {
        this.states.push(state);
    }

    async delayedResolve(): Promise<unknown[]> {
        await wait();
        return this.states;
    }

    getStates(): unknown[] {
        return this.states;
    }

    getLatestState(): unknown {
        return this.states[this.states.length - 1];
    }

    clear(): void {
        this.states.length = 0;
    }

    async renderToSnapshot(component: ReactNode): Promise<void> {
        render(component);
        expect(await this.delayedResolve()).toMatchSnapshot();
    }
}
