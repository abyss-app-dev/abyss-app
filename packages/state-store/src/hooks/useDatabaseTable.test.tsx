import { buildTestDB, type SQliteClient, SqliteTable } from '@abyss/records';
import { useEffect, useState } from 'react';
import { beforeEach, describe, test } from 'vitest';
import { DatabaseProvider } from '../context/DatabaseContext';
import { StateRecorder } from '../mocking/state-recorder';
import { TestRecorder } from '../mocking/wrapped-component';
import { useDatabaseTable } from './useDatabaseTable';

describe('useDatabaseTable hook', () => {
    let db: SQliteClient;
    let recorder: StateRecorder;

    beforeEach(async () => {
        db = await buildTestDB();
        recorder = new StateRecorder();
    });

    test('loads table instance', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const result = useDatabaseTable(SqliteTable.settings);
                        return {
                            loading: result.loading,
                            error: result.error,
                            hasTable: !!result.data,
                            tableName: result.data?.constructor.name,
                        };
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('manual refetch functionality', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [hasRefetched, setHasRefetched] = useState(false);

                        const result = useDatabaseTable(SqliteTable.settings);

                        useEffect(() => {
                            if (result.data && !result.loading && !hasRefetched) {
                                setHasRefetched(true);
                                setTimeout(() => result.refetch());
                            }
                        }, [result.data, result.loading, result.refetch, hasRefetched]);

                        return {
                            loading: result.loading,
                            error: result.error,
                            hasTable: !!result.data,
                            hasRefetched,
                        };
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('subscription receives table updates', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [hasSetData, setHasSetData] = useState(false);

                        const result = useDatabaseTable(SqliteTable.settings);

                        useEffect(() => {
                            if (result.data && !result.loading && !hasSetData && result.setData) {
                                setHasSetData(true);
                                setTimeout(() => result.setData(result.data));
                            }
                        }, [result.data, result.loading, result.setData, hasSetData]);

                        return {
                            loading: result.loading,
                            error: result.error,
                            hasTable: !!result.data,
                            hasSetData,
                        };
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('handles different table types', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const settingsResult = useDatabaseTable(SqliteTable.settings);

                        return {
                            settingsLoading: settingsResult.loading,
                            settingsError: settingsResult.error,
                            hasSettingsTable: !!settingsResult.data,
                            settingsTableName: settingsResult.data?.constructor.name,
                        };
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('table subscription persists across re-renders', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [renderCount, setRenderCount] = useState(1);

                        const result = useDatabaseTable(SqliteTable.settings);

                        useEffect(() => {
                            if (renderCount < 3) {
                                setTimeout(() => setRenderCount(prev => prev + 1));
                            }
                        }, [renderCount]);

                        return {
                            loading: result.loading,
                            error: result.error,
                            hasTable: !!result.data,
                            renderCount,
                        };
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('subscription updates when database table changes', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [hasTriggeredChange, setHasTriggeredChange] = useState(false);
                        const [refetchCount, setRefetchCount] = useState(0);

                        const result = useDatabaseTable(SqliteTable.settings);

                        // Track refetch calls by monitoring data changes
                        useEffect(() => {
                            if (result.data && !result.loading) {
                                setRefetchCount(prev => prev + 1);
                            }
                        }, [result.data, result.loading]);

                        // Trigger a database change that should cause subscription update
                        useEffect(() => {
                            if (result.data && !result.loading && !hasTriggeredChange) {
                                setHasTriggeredChange(true);
                                setTimeout(async () => {
                                    // Creating a new record should trigger table subscription
                                    await db.tables[SqliteTable.settings].default();
                                }, 100);
                            }
                        }, [result.data, result.loading, hasTriggeredChange]);

                        return {
                            loading: result.loading,
                            error: result.error,
                            hasTable: !!result.data,
                            hasTriggeredChange,
                            refetchCount,
                            dataLength: Array.isArray(result.data) ? result.data.length : 0,
                        };
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('multiple table subscriptions work independently', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [changeStep, setChangeStep] = useState(0);

                        const settings1 = useDatabaseTable(SqliteTable.settings);
                        const settings2 = useDatabaseTable(SqliteTable.settings);

                        // Trigger database operations
                        useEffect(() => {
                            if (settings1.data && settings2.data && !settings1.loading && !settings2.loading && changeStep === 0) {
                                setChangeStep(1);
                                setTimeout(async () => {
                                    // Both subscriptions should receive this update
                                    await db.tables[SqliteTable.settings].default();
                                    setChangeStep(2);
                                }, 100);
                            }
                        }, [settings1.data, settings1.loading, settings2.data, settings2.loading, changeStep]);

                        return {
                            settings1Loading: settings1.loading,
                            settings2Loading: settings2.loading,
                            settings1HasData: !!settings1.data,
                            settings2HasData: !!settings2.data,
                            changeStep,
                            bothReady: !!settings1.data && !!settings2.data && !settings1.loading && !settings2.loading,
                        };
                    }}
                />
            </DatabaseProvider>
        );
    });
});
