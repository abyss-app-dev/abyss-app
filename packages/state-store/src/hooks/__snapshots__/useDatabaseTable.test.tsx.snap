// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useDatabaseTable hook > handles different table types 1`] = `
[
  {
    "hasSettingsTable": false,
    "settingsError": null,
    "settingsLoading": true,
    "settingsTableName": undefined,
  },
  {
    "hasSettingsTable": true,
    "settingsError": null,
    "settingsLoading": false,
    "settingsTableName": "Array",
  },
]
`;

exports[`useDatabaseTable hook > loads table instance 1`] = `
[
  {
    "error": null,
    "hasTable": false,
    "loading": true,
    "tableName": undefined,
  },
  {
    "error": null,
    "hasTable": true,
    "loading": false,
    "tableName": "Array",
  },
]
`;

exports[`useDatabaseTable hook > manual refetch functionality 1`] = `
[
  {
    "error": null,
    "hasRefetched": false,
    "hasTable": false,
    "loading": true,
  },
  {
    "error": null,
    "hasRefetched": false,
    "hasTable": true,
    "loading": false,
  },
  {
    "error": null,
    "hasRefetched": true,
    "hasTable": true,
    "loading": false,
  },
  {
    "error": null,
    "hasRefetched": true,
    "hasTable": true,
    "loading": true,
  },
  {
    "error": null,
    "hasRefetched": true,
    "hasTable": true,
    "loading": false,
  },
]
`;

exports[`useDatabaseTable hook > multiple table subscriptions work independently 1`] = `
[
  {
    "bothReady": false,
    "changeStep": 0,
    "settings1HasData": false,
    "settings1Loading": true,
    "settings2HasData": false,
    "settings2Loading": true,
  },
  {
    "bothReady": false,
    "changeStep": 0,
    "settings1HasData": true,
    "settings1Loading": false,
    "settings2HasData": false,
    "settings2Loading": true,
  },
  {
    "bothReady": true,
    "changeStep": 0,
    "settings1HasData": true,
    "settings1Loading": false,
    "settings2HasData": true,
    "settings2Loading": false,
  },
  {
    "bothReady": true,
    "changeStep": 1,
    "settings1HasData": true,
    "settings1Loading": false,
    "settings2HasData": true,
    "settings2Loading": false,
  },
]
`;

exports[`useDatabaseTable hook > subscription receives table updates 1`] = `
[
  {
    "error": null,
    "hasSetData": false,
    "hasTable": false,
    "loading": true,
  },
  {
    "error": null,
    "hasSetData": false,
    "hasTable": true,
    "loading": false,
  },
  {
    "error": null,
    "hasSetData": true,
    "hasTable": true,
    "loading": false,
  },
  {
    "error": null,
    "hasSetData": true,
    "hasTable": true,
    "loading": false,
  },
]
`;

exports[`useDatabaseTable hook > subscription updates when database table changes 1`] = `
[
  {
    "dataLength": 0,
    "error": null,
    "hasTable": false,
    "hasTriggeredChange": false,
    "loading": true,
    "refetchCount": 0,
  },
  {
    "dataLength": 0,
    "error": null,
    "hasTable": true,
    "hasTriggeredChange": false,
    "loading": false,
    "refetchCount": 0,
  },
  {
    "dataLength": 0,
    "error": null,
    "hasTable": true,
    "hasTriggeredChange": true,
    "loading": false,
    "refetchCount": 1,
  },
]
`;

exports[`useDatabaseTable hook > table subscription persists across re-renders 1`] = `
[
  {
    "error": null,
    "hasTable": false,
    "loading": true,
    "renderCount": 1,
  },
  {
    "error": null,
    "hasTable": false,
    "loading": true,
    "renderCount": 2,
  },
  {
    "error": null,
    "hasTable": true,
    "loading": false,
    "renderCount": 2,
  },
  {
    "error": null,
    "hasTable": true,
    "loading": false,
    "renderCount": 3,
  },
]
`;
